fn/mvar.go

```diff
@@ -0,0 +1,205 @@
+package fn
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// MVar[A any] is a structure that is designed to store a single value in an API
+// that dispenses with data races. Think of it as a box for a value.
+//
+// It has two states: full and empty.
+//
+// It supports two operations: take and put.
+//
+// The state transition rules are as follows:
+//  1. put while full blocks.
+//  2. put while empty sets.
+//  3. take while full resets.
+//  4. take while empty blocks.
+//  5. read while full nops.
+//  6. read while empty blocks.
+type MVar[A any] struct {
```

Comment:
Why do we need this when we already have [`atomic`](https://pkg.go.dev/sync/atomic) and [`sync`](https://pkg.go.dev/sync)? Or what are the use cases?

Reply:
The use case is not having to think about synchronization as a separate thing from state management. If you need shared state across threads you can use one of these instead of rolling your own every time. It's not quite as good as STM but I don't have the skill/knowledge to quickly implement a nice STM API at the moment.

In my case here I made this because when I was designing the rest of this change I reached for [this API](https://hackage.haskell.org/package/base-4.19.0.0/docs/Control-Concurrent-MVar.html) that is part of the standard library in Haskell 

I personally don't think anyone should be using the usual sync primitives directly in application code. It's error prone and there are better abstractions to use out there such as this one. I certainly don't want to think about sync primitives when I'm writing domain specific logic.